tests/integration/src/xcm_api.rs

```diff
@@ -188,3 +198,55 @@ fn query_delivery_fees_is_ok() {
         );
     })
 }
+
+#[test]
+fn dry_run_call_is_ok() {
+    new_test_ext().execute_with(|| {
+        // Prepare a dummy origin and call
+        let origin = OriginCaller::system(frame_system::RawOrigin::Root.into());
+        let call = RuntimeCall::System(frame_system::Call::remark {
+            remark: vec![0u8; 32],
+        });
+
+        let result = Runtime::dry_run_call(origin, call).expect("Must return some effects.");
+        assert_eq!(result.forwarded_xcms.len(), 1);
+    })
+}
+
+// Minimal test in a runtime isolated environment to make sure `dry_run_xcm` is callable
+#[test]
+fn dry_run_xcm_is_ok() {
+    new_test_ext().execute_with(|| {
+        // Prepare a dummy location and xcm sequence
+        let location = VersionedLocation::V4(Location::here());
+        let xcm_sequence = Xcm::<()>::builder_unsafe()
+            .clear_error()
+            .unsubscribe_version()
+            .build();
+        let dummy_message =
+            Xcm::<RuntimeCall>::from(VersionedXcm::V4(xcm_sequence).try_into().unwrap());
+        let versioned_xcm = VersionedXcm::V4(dummy_message);
+
+        let result =
+            Runtime::dry_run_xcm(location, versioned_xcm).expect("Must return some effects.");
+        assert_eq!(result.forwarded_xcms.len(), 1);
+    })
```

Comment:
I should have been more explicit with my comment before - the question/suggestion was related to this test in particular. It passed even though empty implementation was used. Which makes it dubious of the tests usefulness, right?

What I'd suggest to do is to improve this test by sending it different XCM sequence, one that ideally has some more side effects.

The added test below, `xcm_recorder_configuration_is_ok` seems more like a _unit test_ then an integration test, IMHO.

Reply:
I've made improvements to the `dry_run_xcm` test by using a more realistic XCM sequence. Now, it provides a complete execution result for all our runtime XCM configurations. However, since `dry_run_xcm` doesn't interact with the `XcmRecorder`, the test still passes even with a default implementation.

Next, I applied the same approach to `dry_run_call`, which does use the `XcmRecorder` and seemed like the ideal scenario. However, I ran into a persistent **"SendFailure"** error when trying to execute the `limited_reserve_transfer_assets` XCM call. No matter how I configure the call, the error remains, and the documentation hasn’t been particularly helpful: https://github.com/paritytech/polkadot-sdk/blob/51f336711a0391987db69d6281c9b57bfe49d925/polkadot/xcm/pallet-xcm/src/lib.rs#L524-L526.
Maybe it is because in a simulated environment? Do you have any alternative ideas? I’ve kept my original test at the bottom of the file as reference point.

Here's what I propose:
1. Preserve the original `dry_run_call` test using a remark call. This will at least confirm that the `DryRunApi` is properly implemented in the runtimes.
2. Keep `xcm_recorder_configuration_is_ok` to check whether the `XcmRecorder` is correctly configured in the runtime XCM configs. While this test seems weak, it still can be considered as a integration test ensuring the correct configuration of the `XcmRecorder`.

WDYT?

Reply:
Debugging those XCM issues can be very time consuming 😓.
I'd suggest to try and force the used XCM version of the destination you're sending the message to.

Your suggestions look good to me, let's just keep it as simple as possible while testing that the functionality actually works. Please don't spend too much time on this anymore 🙈

Reply:
I quickly tested again with `force_xcm_version` and `force_default_xcm_version`, but again, same error message.
I deleted the commented ideal test and replaced it with a small comment for posterity. If you agree, I think we can mark this as resolved and once approved, merge this PR.

Reply:
Yes, no need to dwell on this anymore 👍